src/Common/src/Interop/Unix/System.Native/Interop.FdSet.cs

```diff
@@ -2,34 +2,54 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Diagnostics;
 using System.Runtime.InteropServices;
 
 internal static partial class Interop
 {
     internal static partial class Sys
     {
-        public unsafe struct FdSet
-        {
-            const int FDSET_MAX_FDS = 1024;
-            const int FDSET_NFD_BITS = FDSET_MAX_FDS / (8 * sizeof(uint));
+        [DllImport(Libraries.SystemNative)]
+        private static extern int FdSetSize();
+
+        private static readonly int FD_SETSIZE_BITS = FdSetSize();
+
+        private const int BitsPerByte = 8;
+        private static readonly int FD_SETSIZE_BYTES = FD_SETSIZE_BITS / BitsPerByte;
+
+        internal static readonly int FD_SETSIZE_UINTS = FD_SETSIZE_BYTES / sizeof(uint);
 
-            private fixed uint _bits[FDSET_MAX_FDS / FDSET_NFD_BITS];
-  
-            public void Set(int fd)
+        internal static unsafe void FD_SET(int fd, uint* fdset)
+        {
+            if (fd >= FD_SETSIZE_BITS)
             {
-                fixed (uint* bits = _bits)
-                {
-                    bits[fd / FDSET_NFD_BITS] |= (1u << (fd % FDSET_NFD_BITS));
-                }
+                ThrowInvalidFileDescriptor(fd);
             }
+            Debug.Assert(fdset != null);
+
+            fdset[fd / FD_SETSIZE_UINTS] |= (1u << (fd % FD_SETSIZE_UINTS));
+        }
 
-            public bool IsSet(int fd)
+        internal static unsafe bool FD_ISSET(int fd, uint* fdset)
+        {
+            if (fd >= FD_SETSIZE_BITS)
             {
-                fixed (uint* bits = _bits)
-                {
-                    return (bits[fd / FDSET_NFD_BITS] & (1U << (fd % FDSET_NFD_BITS))) != 0;
-                }
+                ThrowInvalidFileDescriptor(fd);
             }
+            Debug.Assert(fdset != null);
+
+            return (fdset[fd / FD_SETSIZE_UINTS] & (1u << (fd % FD_SETSIZE_UINTS))) != 0;
+        }
+
+        internal static unsafe void FD_ZERO(uint* fdset)
+        {
+            Debug.Assert(fdset != null);
+            Interop.Sys.MemSet(fdset, 0, (UIntPtr)FD_SETSIZE_BYTES);
```

Comment:
Developers can change it to increase the scale of their servers.  This problem was originally discovered because of that exact increase, e.g. a search turns up examples like https://cs.uwaterloo.ca/~brecht/servers/openfiles.html.